stdlib/public/core/UnicodeScalarProperties.swift

```diff
@@ -0,0 +1,1449 @@
+      case __swift_stdlib_U_SURROGATE: self = .surrogate
+      case __swift_stdlib_U_DASH_PUNCTUATION: self = .dashPunctuation
+      case __swift_stdlib_U_START_PUNCTUATION: self = .openPunctuation
+      case __swift_stdlib_U_END_PUNCTUATION: self = .closePunctuation
+      case __swift_stdlib_U_CONNECTOR_PUNCTUATION: self = .connectorPunctuation
+      case __swift_stdlib_U_OTHER_PUNCTUATION: self = .otherPunctuation
+      case __swift_stdlib_U_MATH_SYMBOL: self = .mathSymbol
+      case __swift_stdlib_U_CURRENCY_SYMBOL: self = .currencySymbol
+      case __swift_stdlib_U_MODIFIER_SYMBOL: self = .modifierSymbol
+      case __swift_stdlib_U_OTHER_SYMBOL: self = .otherSymbol
+      case __swift_stdlib_U_INITIAL_PUNCTUATION: self = .initialPunctuation
+      case __swift_stdlib_U_FINAL_PUNCTUATION: self = .finalPunctuation
+      default: fatalError("Unknown general category \(rawValue)")
+      }
+    }
+  }
+}
+
+extension Unicode.Scalar.Properties {
+
+  /// The general category (most usual classification) of the scalar.
+  ///
+  /// This property corresponds to the `General_Category` property in the
+  /// [Unicode Standard](http://www.unicode.org/versions/latest/).
+  public var generalCategory: Unicode.GeneralCategory {
+    let rawValue = __swift_stdlib_UCharCategory(
+      UInt32(__swift_stdlib_u_getIntPropertyValue(
+        _value, __swift_stdlib_UCHAR_GENERAL_CATEGORY)))
+    return Unicode.GeneralCategory(rawValue: rawValue)
+  }
+}
+
+/// Executes a block that will attempt to store text in the memory owned by the
+/// given string storage object, returning the length of the text.
+///
+/// The closure is assumed to return an `Int32` value (the convention of ICU's C
+/// functions) that is equal to the length of the string being stored. If the
+/// storage object is not large enough to hold the string, it is replaced by one
+/// that is exactly the required size and the closure is executed one more time
+/// to populate it.
+@_inlineable
+@_versioned
+internal func _expandingStorageIfNeeded<
+  CodeUnit: UnsignedInteger & FixedWidthInteger
+>(
+  _ storage: inout _SwiftStringStorage<CodeUnit>,
+  body: (_SwiftStringStorage<CodeUnit>) throws -> Int32
+) rethrows -> Int {
+  let z = try body(storage)
+  let correctSize = Int(z)
+  if correctSize > storage.capacity {
+    // If the buffer wasn't large enough, replace it with one that is the
+    // correct size and execute the closure again.
+    storage = _SwiftStringStorage<CodeUnit>.create(
+      capacity: correctSize,
+      count: correctSize)
+    return Int(try body(storage))
+  } else {
+    // Otherwise, the buffer has been populated; update its count to the correct
+    // value.
+    storage.count = correctSize
+    return correctSize
+  }
+}
+
+extension Unicode.Scalar.Properties {
+
+  internal func _scalarName(
+    _ choice: __swift_stdlib_UCharNameChoice
+  ) -> String? {
+    let initialCapacity = 256
+
+    var storage = _SwiftStringStorage<UTF8.CodeUnit>.create(
```

Comment:
Use `String._fromUTF8CodeUnitSequence/_fromWellFormedUTF8CodeUnitSequence/_fromASCI/`, which will attempt to form a small string if possible. Try to avoid creating a storage class.

You could make a small string and try forming into that, falling back to an Array+`_fromASCII` otherwise. See e.g. `_SmallString._withMutableExcessCapacityBytes`. Micro-optimizing further is almost certainly not worth it here.

Reply:
The good news is that this code doesn't compile anymore, so it's a great time to fix it!

```swift
/Users/buildnode/jenkins/workspace-private/swift-PR-compiler-performance-macOS/master/new/swift/stdlib/public/core/UnicodeScalarProperties.swift:1091:12: error: argument labels '(_storage:)' do not match any available overloads
16:35:05     return String(_storage: storage)
```

Reply:
I have a fix for case conversion (that uses a stack buffer). I'll try to do _scalarName too and share that

Reply:
I've fixed `_scalarName` and pushed it—thanks for the pointers to the new APIs.

I hadn't gotten to case conversion yet since those need to be written into UTF-16 buffers and I didn't see a small analogue of `_SmallUTF8String`, so I figured there would be some extra work that I haven't been able to dive into yet. If you already have a fix, I'll just wait for that!